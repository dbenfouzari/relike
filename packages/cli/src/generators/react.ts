import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
