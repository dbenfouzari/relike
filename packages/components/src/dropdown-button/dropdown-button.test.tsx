import { fireEvent, render, screen } from "@testing-library/react";
import classNames from "classnames";

import DropdownButton, { getMaxChildrenWidth } from "./dropdown-button";
import classes from "./dropdown-button.module.scss";

/**
 * Helper to generate dropdown items.
 *
 * @param   {number}                            [n = 100] Number of items to generate
 * @example
 * generateItems() // generates 100 items
 * generateItems(5) // generates 5 items
 * @returns {{ value: number, text: string }[]}           The items.
 */
const generateItems = (n = 5) => {
  const result = [];
  for (let i = 0; i < n; i++) {
    result.push({ value: i.toString(), text: `Item ${i}` });
  }
  return result;
};

describe("DropdownButton", () => {
  it("should render successfully", () => {
    render(<DropdownButton items={generateItems()} />);

    expect(screen.getByTestId("dropdown-button")).toMatchSnapshot();
  });

  it("should render successfully with initial value", () => {
    render(<DropdownButton items={generateItems()} value="2" />);

    expect(screen.getByTestId("dropdown-button")).toMatchSnapshot();
  });

  it("should open", () => {
    render(<DropdownButton items={generateItems()} />);
    const button = screen.getByTestId("dropdown-button");

    fireEvent.click(button);

    const elm = screen.getByTestId(2);

    fireEvent.click(elm);
  });

  it("should render successfully when allow empty value", () => {
    render(<DropdownButton items={generateItems()} allowEmpty />);

    expect(screen.getByTestId("empty-placeholder")).toBeInTheDocument();
  });

  it("should handle selection", () => {
    const onSelect = jest.fn();

    render(<DropdownButton items={generateItems()} onChange={onSelect} />);
    const button = screen.getByTestId("dropdown-button");

    fireEvent.click(button);
    fireEvent.click(screen.getByTestId("button-4"));

    expect(onSelect).toHaveBeenCalledWith("4");
  });

  it("should close on click outside", () => {
    /**
     * Dumb component, just to be able to click outside the dropdown button
     *
     * @returns {JSX.Element} The DropdownButton component
     * @example
     * <Component />
     */
    const Component = () => (
      <div data-testid="dumb-wrapper">
        <DropdownButton items={generateItems()} />
      </div>
    );

    render(<Component />);

    const button = screen.getByTestId("dropdown-button");

    // Dropdown button should NOT be open
    expect(button.className).not.toEqual(classNames(classes.wrapper, classes.wrapper__open));

    fireEvent.click(button);
    // Dropdown button should be open
    expect(button.className).toEqual(classNames(classes.wrapper, classes.wrapper__open));
    // Click outside the dropdown button
    fireEvent.click(screen.getByTestId("dumb-wrapper"));
    // Dropdown button should NOT be open
    expect(button.className).not.toEqual(classNames(classes.wrapper, classes.wrapper__open));
  });

  it("should show value label when value is passed", () => {
    render(<DropdownButton items={generateItems()} value="2" />);

    expect(screen.getAllByText("Item 2")).toHaveLength(2);
  });

  it("getMaxChildrenWidth should return nothing when no parent given", () => {
    expect(getMaxChildrenWidth(null)).toBe(undefined);
  });
});
