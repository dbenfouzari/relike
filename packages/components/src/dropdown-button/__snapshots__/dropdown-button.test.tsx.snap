// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownButton should render successfully 1`] = `
<button
  class="wrapper"
  data-testid="dropdown-button"
>
  <span
    style="min-width: 0;"
  >
    <span
      class="value_placeholder"
    />
    <span
      class="jsx-786785084 icon material-icons"
    >
      arrow_drop_down
    </span>
  </span>
  <ul
    class="item_list"
  >
    <li
      class="item_list_item"
      data-testid="0"
      id="dropdown-value-0"
    >
      <button
        data-testid="button-0"
      >
        Item 0
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="1"
      id="dropdown-value-1"
    >
      <button
        data-testid="button-1"
      >
        Item 1
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="2"
      id="dropdown-value-2"
    >
      <button
        data-testid="button-2"
      >
        Item 2
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="3"
      id="dropdown-value-3"
    >
      <button
        data-testid="button-3"
      >
        Item 3
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="4"
      id="dropdown-value-4"
    >
      <button
        data-testid="button-4"
      >
        Item 4
      </button>
    </li>
  </ul>
</button>
`;

exports[`DropdownButton should render successfully with initial value 1`] = `
<button
  class="wrapper"
  data-testid="dropdown-button"
>
  <span
    style="min-width: 0;"
  >
    <span>
      Item 2
    </span>
    <span
      class="jsx-786785084 icon material-icons"
    >
      arrow_drop_down
    </span>
  </span>
  <ul
    class="item_list"
  >
    <li
      class="item_list_item"
      data-testid="0"
      id="dropdown-value-0"
    >
      <button
        data-testid="button-0"
      >
        Item 0
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="1"
      id="dropdown-value-1"
    >
      <button
        data-testid="button-1"
      >
        Item 1
      </button>
    </li>
    <li
      class="item_list_item item_list_item__selected"
      data-testid="2"
      id="dropdown-value-2"
    >
      <button
        data-testid="button-2"
      >
        Item 2
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="3"
      id="dropdown-value-3"
    >
      <button
        data-testid="button-3"
      >
        Item 3
      </button>
    </li>
    <li
      class="item_list_item"
      data-testid="4"
      id="dropdown-value-4"
    >
      <button
        data-testid="button-4"
      >
        Item 4
      </button>
    </li>
  </ul>
</button>
`;
